python-package/lightgbm/basic.py

```diff
@@ -594,6 +599,79 @@ def _load_pandas_categorical(file_name=None, model_str=None):
         return None
 
 
+class Sequence(object):
+    """
+    Generic data access interface.
+
+    Object should support the following operations:
+
+        # Get total row number.
+        >>> len(seq)
+        # Random access by row index. Use for data sampling.
+        >>> seq[10]
+        # Range data access. Use to read data in batch when constructing Dataset.
+        >>> seq[0:100]
+        # Optionally specify batch_size to control range data read size.
+        >>> seq.batch_size
+
+    With random access, data sampling does not need to go through all data.
+    With range data access, there's no need to read all data into memory thus
+    reduce memory usage.
+    """
+
+    __metaclass__ = abc.ABCMeta
+
+    batch_size = 4096  # Defaults to read 4K rows in each batch.
+
+    @staticmethod
+    def is_class(obj):
+        """Check if objection is instance of class Sequence.
+
+        Args:
+        -------
+            obj ([any]): object to be checked
+
+        Returns
+        -------
+            [bool]: is Sequence class
+        """
+        if isinstance(obj, list) or hasattr(obj, "getformat"):
+            return False
+        return hasattr(obj, "__getitem__") and hasattr(obj, "__len__")
+
+    @abc.abstractmethod
+    def __getitem__(self, idx):  # type: (Union[int, slice]) -> np.ndarray
+        """Return data for given row index.
+
+        A basic implementation should look like this:
+
+        .. code-block:: python
+
+            if isinstance(idx, numbers.Integral):
+                return self.__get_one_line__(idx)
+            elif isinstance(idx, slice):
+                return np.stack(self.__get_one_line__(i) for i in range(idx.start, idx.stop))
+            else:
+                raise TypeError("Sequence index must be integer or slice, got {}".format(type(idx)))
+
+        Parameters
+        ----------
+        idx : int, slice[int]
+            Item index.
+
+        Returns
+        -------
+        result : numpy 1-D array, numpy 2-D array
+            1-D array if idx is int, 2-D array if idx is slice
+        """
+        raise NotImplementedError("remove this line if subclassing")
+
+    @abc.abstractmethod
+    def __len__(self):  # type: () -> int
```

Comment:
```suggestion
    def __len__(self) -> int:
```

Is there a reason that you prefer to use type hint comments instead of type hints in code? As of https://github.com/microsoft/LightGBM/pull/3581 this project no longer supports Python 2.x, so you can use type hints in code directly.

In fact, based on #3756 we have a preference for it.

Reply:
Unless there is a specific reason to use comments, can you please remove them in this PR and replace them with type hints in code? For consistency with the rest of the project.

Reply:
It's for Python 2.x support. It's a pleasure to drop Python 2.x support.

Reply:
Will do +1